[ConstraintSolver] Fix handling of defaultable constraint in `getPotentialBindings`

xedin opened the pull request:
While trying to find a fixed type for a given type variable, check if
it has representative and if it does, reflect that in the returned type.

Resolves: rdar://problem/34670592

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

slavapestov commented on lib/Sema/CSBindings.cpp:
```diff
@@ -229,8 +229,8 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
       // Do these in a separate pass.
-      if (getFixedTypeRecursive(constraint->getFirstType(), true)
-              ->getAs<TypeVariableType>() == typeVar) {
+      if (simplifyType(constraint->getFirstType())->getAs<TypeVariableType>() ==
```
Is this part necessary?

xedin replied:
It looks like if I make `getFixedTypeRecursive` actually return representative it crashes trying to apply arguments in `applySolution`. I'm going to take a closer look at that, so maybe `simplifyType` is not actually required here, but it seems like `getFixedTypeRecursive` was not intended to return representative type variable...

xedin committed:
[ConstraintSolver] Fix handling of defaultable constraint in `getPotentialBindings`

While trying to find a fixed type for a given type variable, check if
it has representative and if it does, reflect that in the returned type.

Resolves: rdar://problem/34670592

slavapestov commented on test/Constraints/diagnostics.swift:
```diff
@@ -1111,3 +1111,6 @@ func elephant<T, U>(_: T) where T : Collection, T.Element == U, T.Element : Hash
     _ = elephant(a) // expected-error {{generic parameter 'U' could not be inferred}}
 }
+
+// rdar://problem/34670592 - Compiler crash on heterogeneous collection literal
```
This is not a diagnostics test since it does not diagnose, perhaps there is a better place we can put it?

xedin replied:
I was just thinking about that, moved it to `test/Constraints/construction.swift`

slavapestov commented:
Nice!

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@slavapestov If you are okay with this I'm going to merge.
